sdk/storage/AzureStorageBlob/Source/DataStructures/StorageStreamUploader.swift

```diff
@@ -0,0 +1,540 @@
+    public var progress = 0
+
+    /// The list of blocks for the blob upload.
+    public var blockList = [(range: Range<Int>, blockId: String)]()
+
+    /// Internal list that maps the order of block IDs
+    internal var blockIdMap = [String: Int]()
+
+    /// Logs completed block IDs and the order they *should* be in
+    internal var completedBlockMap = [String: Int]()
+
+    /// Indicates if the upload  is complete.
+    public var isComplete: Bool {
+        return progress == fileSize
+    }
+
+    /// Indicates if the upload  is encrypted.
+    public var isEncrypted: Bool {
+        return options.encryptionOptions?.key != nil || options.encryptionOptions?.keyResolver != nil
+    }
+
+    internal let client: StorageBlobClient
+
+    internal let options: UploadBlobOptions
+
+    /// Size, in bytes, of the file to be uploaded.
+    internal var fileSize: Int!
+
+    // MARK: Initializers
+
+    /// Create a `BlobStreamUploader` object.
+    /// - Parameters:
+    ///   - client: A`StorageBlobClient` reference.
+    ///   - name: The name of the blob to upload.
+    ///   - container: The name of the container the blob is contained in.
+    ///   - options: An `UploadBlobOptions` object to control the upload process.
+    public init(
+        client: StorageBlobClient,
+        delegate: BlobUploadDelegate? = nil,
+        source: URL,
+        name: String,
+        container: String,
+        properties: BlobProperties? = nil,
+        options: UploadBlobOptions? = nil
+    ) throws {
+        let manager = FileManager.default
+        guard manager.fileExists(atPath: source.path) else {
+            throw AzureError.general("File not found: \(source.path)")
+        }
+        let attributes = try manager.attributesOfItem(atPath: source.path)
+        guard let fileSize = attributes[FileAttributeKey.size] as? Int else {
+            throw AzureError.general("Unable to determine file size: \(source.path)")
+        }
+        self.fileSize = fileSize
+
+        self.client = client
+        self.delegate = delegate
+        self.options = options ?? UploadBlobOptions()
+        self.blobName = name
+        self.containerName = container
+        self.uploadSource = source
+        self.blobProperties = properties
+        self.blockList = computeBlockList()
+    }
+
+    // MARK: Public Methods
+
+    /// Read and return the content of the downloaded file.
+    public func contents() throws -> Data {
+        let handle = try FileHandle(forReadingFrom: uploadSource)
+        defer { handle.closeFile() }
+        return handle.readDataToEndOfFile()
+    }
+
+    /// Uploads the entire blob in a parallel fashion.
+    /// - Parameters:
+    ///   - group: An optional `DispatchGroup` to wait for the download to complete.
+    ///   - completion: A completion handler called when the download completes.
+    public func complete(inGroup group: DispatchGroup? = nil, then completion: @escaping () -> Void) throws {
```

Comment:
If we never store the value of the `completion` parameter, does it need to be marked `@escaping`? I admit I don't know if I fully understand the do's and don'ts of `@escaping`, but I -think- that it's only needed if the lifetime of the parameter outlives that of the method execution.

Reply:
Also applies to other uses of `@escaping` below

Reply:
Short answer is yes. If you remove it, you get the following:
`Escaping closure captures non-escaping parameter 'completion'`

The issue is that the completion closure needs to survive the lifetime of the block that called it. As soon as the async method starts, the original block exits, and thus the completion closure must be marked `@escaping` or it will be released when the calling block finishes.

Reply:
Ah, okay I think I understand that. And having a compile warning when it's needed also very much helps resolve that confusion too :)